### PR TITLE
Using Docker to create and run server images.

### PR DESCRIPTION
**Pull request by icydee**
Scripts to build a docker image, and to run the image on any system which has already installed the docker system (linux, OS X, or Windows).

This should speed up the time to set up a development environment and hopefully make it easier to deploy.

### CODE REVIEW

**Comment by dmcbride** [comment]
Following the instructions as given, I have a few comments:

1. I don't see instructions to *start* the data container (tle-mysql-data).  So I'm assuming that's an oversight.  I'm not sure what flags might be required, especially if there is a socket file required in the containers to be visible.

2. Running "mysql" inside the prebuilt container fails predictably with `ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock' (2)` - but I'm not sure what you're intending here.

3. I'm not sure about the schedulers - ships landing, buildings completing, etc.  I do have a start script that may work here.  However, it'd be really sweet if we got the crontab files into git so that the container could run those as well :)  Anyway, I also have a question about whether beanstalk or memcached are running here (I don't think they are, but I'm not sure).

4. lacuna.conf?  Ideally any lacuna.conf we have in this container would not be visible to git, so we may need to add that to .gitignore.  In my environment, I have /data/Lacuna-Server/etc as a real directory with my local conf file(s) instead of symlinks like I do for bin, lib, and others.  Also, any instructions on how to set up the DBI port for mysql, if required.

File: docker_run.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#
+# You can use GIT as normal in your host to commit your changes.
+#
+# If you 'exit' this container then your server will stop and all data
+# in the container will be lost. EXCEPT for the mysql data with is held
+# in a separate data-only container 'tld-mysql-data'
+#
+echo 'This is your TLE server'
+echo 'you should now run the following command to initialize it'
+echo '  $ ./docker_client_start.sh
```
Comment (dmcbride) [inline docker_run.sh]: Some of these lines are missing end-quotes.  (26 and 32)  Also some (30, 35) are missing quotes altogether, inconsistent with, say, line 27.  Finally, this is probably easier done with cat and heredocs:

cat <<INSTRUCTIONS
This is your TLE server
you should run the following command to initialize it
    \$ ./docker_client_start.sh
...
INSTRUCTIONS

**Comment by dmcbride** [comment]
Following the instructions as given, I have a few comments:

1. I don't see instructions to start the data container (tle-mysql-data). If we don't need it actually running, might I suggest saying so explicitly?

2. Running "mysql" inside the prebuilt container fails predictably with `ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock'` If I instead run the commands shown when I start the container, instead of what's in the setup_docker.txt, I get mysql running, and can continue.  So I'd suggest the `Initializing the database` section mention that you still have to run the docker_client_start script before talking about the docker_exec.sh script.  I'm so busy following the instructions that I didn't notice the console instructions as well :)

3. I'm not sure about the schedulers - ships landing, buildings completing, etc. I do have a start script that may work here. However, it'd be really sweet if we got the crontab files into git so that the container could run those as well :)

**Comment by dmcbride** [comment]
Also, the captcha generation fails:

```
Can't locate GD/SecurityImage.pm in @INC (@INC contains: /data/Lacuna-Server/lib /data/apps/lib/perl5/site_perl/5.12.1/x86_64-linux /data/apps/lib/perl5/site_perl/5.12.1 /data/apps/lib/perl5/5.12.1/x86_64-linux /data/apps/lib/perl5/5.12.1 .) at generate_captcha.pl line 5.
```

Looks like the docker image is missing something :)

**Comment by icydee** [comment]
Thanks for the point about he tle-mysql-data container. It does not actually need to run, it just has to be present! (I will improve the documentation).

Thanks also for the pointer to the 'cat' function. My shell scripting needs work!

It's probably best to ignore this pull request now, I am working on another approach that has a better separation of concerns, but the fact that you were able to run the scripts in a docker container is validation of the approach!

I will try to incorporate your points in the next version.

**Comment by icydee** [comment]
let's not merge this pull request. Better one on the way...

**Pull request closed without merging**